snuba/query/processors/prewhere.py

```diff
@@ -25,56 +34,189 @@ class PrewhereProcessor(QueryProcessor):
+
+    def _get_column_name(self, column: str) -> str:
+        return column
+
+    def _update_conditions(
+        self, query: Query, prewhere_conditions: Sequence[Condition]
+    ) -> None:
+        conditions = query.get_conditions()
+        if not conditions:
+            # This should never happen at this point, but for mypy this can be None.
+            return
+
+        query.set_conditions(
+            [cond for cond in conditions if cond not in prewhere_conditions]
+        )
+        query.set_prewhere(prewhere_conditions)
+
+
+class ASTPrewhereProcessor(PrewhereProcessorDelegate[Expression, Column]):
+    """
+    This class MUST not depend on anything from the legacy implementation because,
+    as long as we have both ast and legacy representation coexisting, this class
+    cannot be sure whether the legacy representation has already been modified
+    to process the pre-where conditions.
+    """
+
+    allowed_ast_operators = [OPERATOR_TO_FUNCTION[o] for o in ALLOWED_OPERATORS]
+
+    def _get_prewhere_candidates(
+        self, query: Query, prewhere_keys: Sequence[str]
+    ) -> Sequence[Tuple[Iterable[Column], Expression]]:
+        # Add any condition to PREWHERE if:
+        # - It is a single top-level condition (not OR-nested), and
+        # - Any of its referenced columns are in prewhere_keys
+        ast_condition = query.get_condition_from_ast()
+        return (
             [
-                (
-                    min(
-                        prewhere_keys.index(col) for col in cols if col in prewhere_keys
-                    ),
-                    cond,
+                (get_columns_in_expression(cond), cond)
+                for cond in get_first_level_conditions(ast_condition)
+                if isinstance(cond, FunctionCall)
+                and cond.function_name in self.allowed_ast_operators
+                and any(
+                    col.column_name in prewhere_keys
+                    for col in get_columns_in_expression(cond)
                 )
-                for cols, cond in prewhere_candidates
-            ],
-            key=lambda priority_and_col: priority_and_col[0],
-        )
-        if prewhere_candidates:
-            prewhere_conditions = [cond for _, cond in prewhere_candidates][
-                :max_prewhere_conditions
             ]
-            query.set_conditions(
-                list(filter(lambda cond: cond not in prewhere_conditions, conditions))
-            )
-        query.set_prewhere(prewhere_conditions)
+            if ast_condition
+            else []
+        )
+
+    def _get_column_name(self, column: Column) -> str:
+        return column.column_name
+
+    def _update_conditions(
+        self, query: Query, prewhere_conditions: Sequence[Expression]
+    ) -> None:
+        ast_condition = query.get_condition_from_ast()
+        if not ast_condition:
+            # This should never happen at this point, but for mypy this can be None.
+            return
```

Comment:
If this is just for mypy, why not `assert ast_condition is not None` rather than silently returning

Reply:
What I did for mypy is to add the if block. Indeed I can replace it with an assertion. Will do.